core/src/main/java/org/infinispan/interceptors/impl/SequentialInterceptorChainImpl.java

```diff
@@ -0,0 +1,342 @@
+
+   public SequentialInterceptor findInterceptorExtending(Class<? extends SequentialInterceptor> interceptorClass) {
+      List<SequentialInterceptor> localInterceptors = this.interceptors;
+      for (SequentialInterceptor interceptor : localInterceptors) {
+         boolean isSubclass = interceptorClass.isInstance(interceptor);
+         if (isSubclass) {
+            return interceptor;
+         }
+      }
+      return null;
+   }
+
+   @Override
+   public SequentialInterceptor findInterceptorWithClass(Class interceptorClass) {
+      List<SequentialInterceptor> localInterceptors = this.interceptors;
+      for (SequentialInterceptor interceptor : localInterceptors) {
+         if (interceptorMatches(interceptor, interceptorClass)) {
+            return interceptor;
+         }
+      }
+      return null;
+   }
+
+   public String toString() {
+      StringBuilder sb = new StringBuilder();
+      List<SequentialInterceptor> localInterceptors = this.interceptors;
+      for (SequentialInterceptor interceptor : localInterceptors) {
+         sb.append("\n\t>> ");
+         sb.append(interceptor);
+      }
+      return sb.toString();
+   }
+
+   public boolean containsInstance(SequentialInterceptor interceptor) {
+      List<SequentialInterceptor> localInterceptors = this.interceptors;
+      for (SequentialInterceptor current : localInterceptors) {
+         if (current == interceptor) {
+            return true;
+         }
+      }
+      return false;
+   }
+
+   public boolean containsInterceptorType(Class<? extends SequentialInterceptor> interceptorType) {
+      return containsInterceptorType(interceptorType, false);
+   }
+
+   @Override
+   public boolean containsInterceptorType(Class<? extends SequentialInterceptor> interceptorType,
+                                          boolean alsoMatchSubClasses) {
+      Class<? extends SequentialInterceptor> sequentialInterceptorType = interceptorType;
+      List<SequentialInterceptor> localInterceptors = this.interceptors;
+      for (SequentialInterceptor interceptor : localInterceptors) {
+         Class<? extends SequentialInterceptor> currentInterceptorType = interceptor.getClass();
+         if (alsoMatchSubClasses) {
+            if (interceptorType.isAssignableFrom(currentInterceptorType)) {
+               return true;
+            }
+            if (sequentialInterceptorType != null &&
+                  sequentialInterceptorType.isAssignableFrom(currentInterceptorType)) {
+               return true;
+            }
+         } else {
+            if (interceptorType == currentInterceptorType ||
+                  sequentialInterceptorType == currentInterceptorType) {
+               return true;
+            }
+         }
+      }
+      return false;
+   }
+
+   @Override
+   public List<SequentialInterceptor> getInterceptors() {
+      return interceptors;
+   }
+
+   protected ImmutableListCopy<SequentialInterceptor> immutableListAdd(List<SequentialInterceptor> list,
```

Comment:
not related to this PR, but we should think to move this methods to InfinispanCollections since they are useful and also used in ClusterCacheStatus.